### PR TITLE
Suppress errors in const eval during trait selection v2

### PR DESCRIPTION
**Pull request by b-naber**
This is the second attempt of https://github.com/rust-lang/rust/pull/92653

Significantly reduces the complexity of the previous PR. Hopefully also result in better perf results.

Needs a perf-run.

r? @oli-obk 

cc @lcnr @RalfJung

### CODE REVIEW

**Comment by b-naber** [comment]
Don't understand why CI outputs that fmt failure. I did run tidy with --bless locally.

**Comment by oli-obk** [comment]
The notable thing in the perf regression is that `eval_to_allocation_raw` is now called in `issue-46449 check | incr-unchanged`, when it wasn't called at all before.

File: compiler/rustc_middle/src/mir/interpret/mod.rs [hunk 1]
```diff
@@ -554,6 +641,129 @@ impl<'tcx> TyCtxt<'tcx> {
...
+        let (param_env, id) = key.into_parts();
+        let dedup_const_map = self.dedup_const_map.lock();
+        debug!("dedup_const_map: {:#?}", dedup_const_map);
+        let alloc_map = dedup_const_map.alloc_map.borrow();
+        debug!("alloc_map: {:#?}", alloc_map);
+
+        let dedup_result = alloc_map.get(&id);
+        debug!(?dedup_result);
+
+        match param_env.reveal() {
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/interpret/mod.rs:685]: these matches are not obvious to me. Can you explain them? I'm unsure what the rules are here.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 2]
```diff
@@ -795,12 +804,48 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
...
+                                uv2.shrink(),
+                                obligation.clone(),
+                                root_obligation,
+                                fallback_has_occurred,
+                            );
+                        }
+                    }
+                    _ => {}
+                }
+                tcx.sess.struct_span_err(span, "failed to evaluate the given constant")
```
Comment (oli-obk) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:847]: all cases of this in the test suite are redundant. maybe make it a delay_span_bug?

File: compiler/rustc_middle/src/mir/interpret/mod.rs (hunk 1 above)
Comment (b-naber) [reply compiler/rustc_middle/src/mir/interpret/mod.rs:685]: The rationale for those was as follows:

If we have stored a successful allocation with `Reveal::Selection` then we can deduplicate that result on calls of `dedup_eval_alloc_raw` with both `Reveal::Selection` and `Reveal::UserFacing` (since `Selection` and `UserFacing` only differ in that we keep certain errors silent with `Selection`, but successful allocations are always equivalent). This is also the reason why we can deduplicate from saved calls with `UserFacing`on calls with both `Selection` and `UserFacing`. 

If we try to deduplicate with `Reveal::All` we should be able to deduplicate from saved query results with both `Selection` and `UserFacing`, since `All` will always succeed if either `UserFacing` or `Selection` have succeeded. On the other hand if we have a saved result with `Reveal::All`, it's not guaranteed that query calls with `Reveal::Selection` or `Reveal::UserFacing` succeed, so we can't deduplicate from those.

**Comment by b-naber** [comment]
> The notable thing in the perf regression is that `eval_to_allocation_raw` is now called in `issue-46449 check | incr-unchanged`, when it wasn't called at all before.

Yeah that's weird. The test case also ICEs with a fingerprint error. I'll try to investigate.

File: compiler/rustc_const_eval/src/interpret/eval_context.rs [hunk 3]
```diff
@@ -291,13 +299,6 @@ impl<'tcx> fmt::Display for FrameInfo<'tcx> {
     }
 }
 
-impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> HasDataLayout for InterpCx<'mir, 'tcx, M> {
-    #[inline]
-    fn data_layout(&self) -> &TargetDataLayout {
-        &self.tcx.data_layout
-    }
-}
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/eval_context.rs:299]: Why did you move this? It was deliberately close to the other `Has*` impls for `InterpCx`

**Comment by b-naber** [comment]
So I tried to investigate this, but I'm unsure about how to test the different scenarios? Does `incr-unchanged` just correspond to another `cargo build` call after a previous successful compilation and no further changes? I assume that `incr-full` is just a new build from scratch with incremental on and `full` is a new build without incremental?!

A couple of points I noticed about the perf regression of the perf benchmark (`issue-46449`) from debug output (new build of the benchmark):

The regression is mostly due to the storing of the results in `dedup_const_map` and the fact that this is a small program in which the evaluation of the constants makes up a large part of the running time. 

There are some things we might want to change:

Currently we store both the results of `eval_to_const_value_raw` and `eval_to_allocation_raw` (which is called by `eval_to_const_value_raw` in all cases except for those involving nullary intrinsics), so that might be redudant. I'm not sure in what circumstances we call these two queries, specifically when we call `eval_to_const_value_raw` on some `GlobalId`, how likely is it that we call `eval_to_allocation_raw` on it again with some other `Reveal` value (if this is unlikely we can certainly get rid of storing the result of the `eval_to_allocation_raw` call [here](https://github.com/rust-lang/rust/blob/53c0bab7d38b05bc9401ad28b8c5b81e23588482/compiler/rustc_const_eval/src/const_eval/eval_queries.rs#L254-L262).

Something else we might to consider is that we currently store the complete result in the deduplication map. At the same time we still cache the result of the query (in the query system), this is redundant. It might be sufficient to either only store some information that allows us to more cleverly make use of cached query results or not cache the query results (not sure how much sense this makes). The first option seems to be the more reasonable approach.

File: compiler/rustc_const_eval/src/interpret/eval_context.rs (hunk 3 above)
Comment (b-naber) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:299]: In the previous PR I had to move `ConstEvalErr` (and some other structs that depend on it) to `rustc_middle`. I pasted this back in the wrong place here, sry. I'll fix this to what it was before.

File: compiler/rustc_middle/src/mir/interpret/mod.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/mir/interpret/mod.rs:685]: 👍 please copy this comment into the source, it explains this perfectly

File: compiler/rustc_middle/src/mir/interpret/mod.rs [hunk 4]
```diff
@@ -554,6 +641,129 @@ impl<'tcx> TyCtxt<'tcx> {
...
+    ) {
+        let dedup_const_map = self.dedup_const_map.lock();
+        dedup_const_map.insert_const_val(id, val);
+        debug!("dedup_const_map after insert: {:#?}", dedup_const_map);
+    }
+
+    /// Tries to deduplicate a call to `eval_to_allocation_raw`. If deduplication isn't
+    /// successful `eval_to_allocation_raw` query is executed.
+    #[instrument(skip(self, opt_span), level = "debug")]
+    pub fn dedup_eval_alloc_raw(
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/interpret/mod.rs:669]: these two functions are almost the same... do you think it would be convoluted to put their bodies into a separate generic function that takes two closures, one for the query and one for the dedup map lookup?

**Comment by oli-obk** [comment]
> Does incr-unchanged just correspond to another cargo build call after a previous successful compilation and no further changes?

I would think so, yes

>I assume that incr-full is just a new build from scratch with incremental on and full is a new build without incremental?!

yes, that's exactly what it is.

> The regression is mostly due to the storing of the results in dedup_const_map and the fact that this is a small program in which the evaluation of the constants makes up a large part of the running time.

That's ok imo, We just need to know *why* there are new queries called, when this PR should only be reducing query calls. As long as we can justify the perf regression, I think we're good, but right now I don't see how that is happening from just the source changes.

**Comment by b-naber** [comment]
The calls of `eval_to_allocation_raw` in the `incr-unchanged` scenario should be fixed now (at least I don't get those calls locally anymore). 

@oli-obk Can you request another perf-run please?

**Comment by b-naber** [comment]
I'll address the review comments after a successful perf-run.

**Comment by lcnr** [comment]
Your approach sadly doesn't work, adding interior mutability to the `TyCtxt` is pretty much always unsound wrt incremental compilation.

For an example how your current approach could break, consider the following scenario:

- initial compilation
    - in `queryA` (e.g. typeck) we evaluate a constant `C` using `dedup_eval_alloc_raw` which calls the query `eval_to_allocation_raw` for `C` with `Reveal::UserFacing`. This adds a dependency edge for `queryA` to `eval_to_allocation_raw`.
    - in `queryB` (e.g. `collect_and_partition_mono_items`) we again evaluate a constant using `dedup_eval_alloc_raw` with `Reveal::All`. `dedup_eval_alloc_raw` reuses the result of the call in `typeck`, meaning that we don't call `eval_to_allocation_raw` in `queryB`. There is therefore no dependency edge from `queryB` to `eval_to_allocation_raw`
- recompilation with changes to body of constant `B`
    - we check whether `queryA` has to be recomputed, see that `eval_to_allocation_raw` for `C` depends on things which changed, and therefore recompute `eval_to_allocation_raw` for `queryA` to check whether it is still the same. Let's assume that the new return value is different. We rerun `queryA` using the new result from `eval_to_allocation_raw` and everything is ok
    - we check whether `queryB` has to be recomputed. All queries `queryB` depends on did not change their output, therefore we don't have to recompute `queryB` which causes subtle bugs and unsoundness. The issue here is that `queryB` **has** to check that `eval_to_allocation_raw` has changed since the last compilation, but because we avoided the query system by using interior mutability, this is not the case.

I believe that the only way to soundly avoid recomputations with different `Reveal`s are the approaches from #81339 and something like #92653 (haven't looked too deeply into that PR). I personally would like to something like #92653 because that also allows us to add much better spans to const eval errors in general, consider something like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=151c1c589fa78413ef194eec9373509b). We should be able to return the eval errors from queries without a significant perf impact, and can probably design an api which makes any misuse (e.g. by forgetting to emit eval errors) very difficult. Am definitely open to talking about this in sync if you're interested.

Sorry for only looking at this PR now and thanks for working on this :bow::heart:

**Comment by b-naber** [comment]
@lcnr Thanks for the review. 

I think the problem you describe should be solved by the changes introduced in the commit `simplify deduplication and better utilize query caching`. Note that we always call the query after these changes and make use of the result cached by the query system. So there should be a dependency edge for `queryB` to `eval_to_allocation_raw`. I could be wrong but, prior to this change we also got an ICE in the query system on the benchmark `issue-46446`, which we don't get anymore after this commit. Have you looked at that commit?

**Comment by lcnr** [comment]
yes, my example was incorrect after this commit, but the issue that queries should remain pure functions still remains. We now get the following scenario:

- initial compilation
    - in `queryA` (e.g. typeck) we evaluate a constant `C` using `dedup_eval_alloc_raw` which calls the query `eval_to_allocation_raw` for `C` with `Reveal::UserFacing`. This adds a dependency edge for `queryA` to `eval_to_allocation_raw`.
    - in `queryB` (e.g. `collect_and_partition_mono_items`) we again evaluate a constant using `dedup_eval_alloc_raw` with `Reveal::All`. `dedup_eval_alloc_raw` sees that `eval_to_allocation_raw` with `Reveal::UserFacing` already succeeded, so we now call that. There is therefore no dependency edge from `queryB` to `eval_to_allocation_raw` with `Reveal::UserFacing`.
- recompilation with changes to body of constant `C`
    - we check whether `queryB` has to be recomputed. As it depends on `eval_to_allocation_raw` with `Reveal::UserFacing`, it gets recomputed and now calls `eval_to_allocation_raw` with `Reveal::All`, as no previous query called it with `Reveal::UserFacing`.

This might seem fine as `eval_to_allocation_raw` with `Reveal::UserFacing` is only used in `queryB` if it's output is the same as with `Reveal::All`, so we never incorrectly mark `queryB` as green.

It still has issues, most noticeably this introduces nondeterminism in the compiler (especially once we run queries in parallel, which we do want). If we first compute `queryB` and then `queryA` then `eval_to_allocation_raw` with `Reveal::All` is called, otherwise not. This can duplicate errors and trigger internal soundness checks - either existing ones or ones added in the future.

I wouldn't be surprised if this causes other subtle issues for our query system, at this point you have to ask people more involved than me

**Comment by b-naber** [comment]
Ok, thanks for clarifying this. I can close this PR, but I would definitely still be interested in working on the issue and getting the idea implemented.

There is one thing I don't understand about your example though:

`QueryB` depends on `eval_to_allocation_raw` with `Reveal::All`, but we never call `eval_to_allocation_raw(Reveal::All)` directly from `queryB`, as this goes through `dedup_eval_alloc`.`dedup_eval_alloc`now calls `eval_to_allocation_raw(Reveal::UserFacing)`, so `queryB` effecively depends on `eval_to_allocation(Reveal::UserFacing)` now, so why does this not cause a dependency edge `queryB` -> `eval_to_allocation_raw(Reveal::UserFacing)` to be created? 

It can't be because the result is already cached, because in a scenario in which we didn't have the deduplication logic and in which some other `queryC` had already computed `eval_to_allocation_raw(Reveal::All)`, we would try to call `eval_to_alloc(Reveal::All)` from queryB, get the cached result, but still create an edge `queryB` -> `eval_alloc(Reveal::All)` in the dependency graph. So why wouldn't we do the analogous thing with the deduplication logic, where `queryB` really depends on `eval_to_alloc(Reveal::UserFacing)`?

**Comment by lcnr** [comment]
that was me not being careful enough while writing it, should have fixed it in the above comment. `queryB` will have an edge to `eval_to_allocation_raw(Reveal::UserFacing)` but not to `eval_to_allocation_raw(Reveal::All)`. I hope the explanation makes more sense now.

> Ok, thanks for clarifying this. I can close this PR, but I would definitely still be interested in working on the issue and getting the idea implemented.

definitely. I would be interested in you reviving #92653. I should be able to take the time for a first look at it this week, either after you're reopened it or even before then if it is already somewhat ready (ignoring perf for now)

**Comment by b-naber** [comment]
Thanks, that makes sense to me now. I'll ping you about further work on this issue on zulip.
